### PR TITLE
Refactor InstructionsCSF

### PR DESCRIPTION
**Pull request by maureensturgeon**
Previously, whether the arrows showed on the CSF instructions was based on both whether the instructions overflowed the content box and whether the instructions were in the collapsed state (must be expanded state to show. arrows). The InstructionsCSF component was very long with a lot of complicated hard-to-parse logic. The changes here 
1. refactor the InstructionsCSF component: split columns into `InstructionsCsfLeftCol`, `InstructionsCsfMiddleCol` and `InstructionsCsfRightCol` components
Screen Shot 2021-04-09 at 10 41 52 AM
2. simplify the logic to calculate height and width of each column
3. update component to show arrows anytime the instructions overflow the container
Screen Shot 2021-04-12 at 10 46 22 AM

## Testing story
-  Added unit tests for new components
- Manually tested CSF levels (such as levels in `s/coursef-2020/stage/4`, for minecraft `/s/coursef-2020/stage/1`)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/templates/instructions/InstructionsCSF.jsx [hunk 1]
```diff
@@ -323,61 +203,43 @@ class InstructionsCSF extends React.Component {
...
+  getMinInstructionsHeight = (collapsed = this.props.collapsed) => {
+    if (collapsed || this.props.overlayVisible || this.props.isEmbedView) {
+      return Math.min(this.getMiddleColHeight(), this.props.maxHeight);
+    } else {
+      return 0;
+    }
+  };
+
+  getMarginsHeight() {
+    const domNode = $(ReactDOM.findDOMNode(this));
```
Comment (cforkish) [inline apps/src/templates/instructions/InstructionsCSF.jsx]: would it be hard to update this to use a ref instead of `findDOMNode`? you could attach it to the outer div returned by `render`...

File: apps/src/templates/instructions/InstructionsCSF.jsx [hunk 2]
```diff
@@ -388,138 +250,65 @@ class InstructionsCSF extends React.Component {
...
-    return this.props.longInstructions && !this.shouldIgnoreShortInstructions();
-  }
+  };
 
-  shouldIgnoreShortInstructions() {
-    // If we have no short instructions, always ignore them.
-    if (!this.props.shortInstructions) {
-      return true;
-    }
+  hasShortInstructions = () => {
```
Comment (cforkish) [inline apps/src/templates/instructions/InstructionsCSF.jsx]: the logic here is a bit confusing and would benefit from comments. some of the ones you removed seem pretty helpful, though they probably need to be updated.

separating out `shortAndLongInstructionsAreEquivalent` into its own function seems useful for giving a name to that part of the logic, but that usefulness is muddied by the checks for short/long instructions at the top of it (the first of which is redundant since you just checked `!!this.props.shortInstructions`, so i'd suggest keeping that part of the logic in this function and having `shortAndLongInstructionsAreEquivalent` only do the levenshtein check. however, since i still think that logic needs a comment despite the helpful function name, at that point i'm not sure it's worth breaking it out into its own function anymore, because the sheer number of helper functions in this component is already a lot to process.

tl;dr i think this would be clearer if you perform the checks for the presence of short/long instructions and return early if possible (with comments), then return the result of the levenshtein check if not (again, with comments)

File: apps/src/templates/instructions/InstructionsCSF.jsx [hunk 3]
```diff
@@ -552,169 +331,37 @@ class InstructionsCSF extends React.Component {
             ],
```
Comment (jamescodeorg) [inline apps/src/templates/instructions/InstructionsCSF.jsx]: Suggestion: move the style defined on L327-332 out of the tag (up into the render function?)

File: apps/src/templates/instructions/InstructionsCsfLeftCol.jsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+const AUTHORED_HINTS_EXTRA_WIDTH = 30; // 40 px, but 10 overlap with prompt icon
+
+const styles = {
+  // bubble has pointer cursor by default. override that if no hints
+  noAuthoredHints: {
+    cursor: 'default',
+    marginBottom: 0
+  },
+  authoredHints: {
+    // raise by 20 so that the lightbulb "floats" without causing the original
```
Comment (jamescodeorg) [inline apps/src/templates/instructions/InstructionsCsfLeftCol.jsx]: Not sure I understand how this comment relates to `marginBottom: 0`?

File: apps/src/templates/instructions/InstructionsCsfLeftCol.jsx [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+  updateDimensions() {
+    this.props.setColWidth(this.getColumnWidth());
+    this.props.setColHeight(this.getColumnHeight());
+  }
+
+  /**
+   * Handle a click to the hint display bubble (lightbulb)
+   */
+  handleClickBubble = () => {
```
Comment (cforkish) [inline apps/src/templates/instructions/InstructionsCsfLeftCol.jsx]: NIT: while you're touching this code, i'd be in favor of renaming this "bubble" to "lightbulb" to disambiguate from progress bubbles

File: apps/src/templates/instructions/InstructionsCSF.jsx [hunk 6]
```diff
@@ -388,138 +250,65 @@ class InstructionsCSF extends React.Component {
...
-    return (
-      !this.shouldIgnoreShortInstructions() &&
-      (this.props.collapsed || !this.props.longInstructions)
-    );
-  }
+  requestHint = () => {
+    this.setState({
+      promptForHint: true
+    });
+    this.props.collapsed && this.props.handleClickCollapser();
```
Comment (jamescodeorg) [inline apps/src/templates/instructions/InstructionsCSF.jsx:314]: Is the intent here to ensure that we are not collapsed?

File: apps/src/templates/instructions/InstructionsCSF.jsx (hunk 2 above)
Comment (maureensturgeon) [reply apps/src/templates/instructions/InstructionsCSF.jsx]: Agreed, this logic is very strange and confusing. I was hoping that if the helper functions were concise and clear that the comments wouldn't be necessary, but it seems like they'd still be valuable so I'll add some back in. I'll also try combining the `shortAndLongInstructionsAreEquivalent ` with `hasShortInstructions` again and see if that clarifies things.

**Review by cforkish — APPROVED** [review]
wowie! this is a really great step in a good direction. this component would clearly benefit from a more comprehensive top-down redesign, but that's obviously out of scope for the task at hand. 

however, as with some of the other refactors we've undertaken, i'm becoming increasingly skeptical of the value of this type of incremental improvement when what's really needed is a more holistic redesign, so i think that's something we should develop some caution around. i realize that sometimes this kind of incremental refactor is necessary simply in order to understand the code well enough to make the needed changes, but in such cases i think it would be wise for us to evaluate the tradeoffs between tackling the bug now vs. deferring for a larger-scoped refactor.

i don't mean to rain on your parade though -- this is great work, and i applaud you for taking it on! i just wanted to call out how this relates to broader issues around our tech debt.

i made a few comments, but nothing blocking.

File: apps/src/templates/instructions/InstructionsCSF.jsx (hunk 2 above)
Comment (cforkish) [reply apps/src/templates/instructions/InstructionsCSF.jsx]: totally. at the very least, preserving the comment about "Levenshtein Edit Distance" would be helpful for those of us who wouldn't automatically know what `levenshtein()` is doing.

not sure if this would work for all of the use cases of this function, but i'm wondering if renaming this to `hasShortAndLongInstructions` and updating the logic to do that full calculation would make this clearer, rather than deferring that to where this is used in the child.

**Review by jamescodeorg — COMMENTED** [review]
I think this is a great step towards making InstructionsCSF more manageable.  I think it would be helpful to have a long header comment on InstructionsCSF that explain the different pieces and how they fit together.

Also, some other things that might be useful to describe in comments somewhere:
- How sizing of the columns works
- How hints work
- How scrolling works
(I think the theme here is that all of these require interactions between the 4 components and that interaction is hard to piece together via the various props, state, and callbacks.)

For future iterations of this component, I'm wondering if it's possible to factor the components a bit more along functional lines (making visual changes if necessary to simplify the code).  It seems like it would be nice if all of the hints logic could be brought together.  Another possibility might be to collect the center and right components into a generic message-displaying component (with scrolling and collapsing) and have hints and instructions "post" messages to that component.

File: apps/src/templates/instructions/InstructionsCsfLeftCol.jsx (hunk 4 above)
Comment (maureensturgeon) [reply apps/src/templates/instructions/InstructionsCsfLeftCol.jsx]: Good catch. This seems like it might be a super old comment that's not relevant anymore. I'll clean up!

File: apps/src/templates/instructions/InstructionsCSF.jsx (hunk 6 above)
Comment (maureensturgeon) [reply apps/src/templates/instructions/InstructionsCSF.jsx:314]: Yeah I believe so, hints are only displayed in the expanded state
